shell/platform/darwin/ios/framework/Source/FlutterViewController.mm

```diff
@@ -1269,39 +1270,56 @@ - (void)startKeyBoardAnimation:(NSTimeInterval)duration {
   }
 
   // Remove running animation when start another animation.
-  // After calling this line,the old display link will invalidate.
   [[self keyboardAnimationView].layer removeAllAnimations];
 
   // Set animation begin value.
   [self keyboardAnimationView].frame =
       CGRectMake(0, _viewportMetrics.physical_view_inset_bottom, 0, 0);
 
-  // Invalidate old display link if the old animation is not complete
-  [self invalidateDisplayLink];
-
-  self.displayLink = [CADisplayLink displayLinkWithTarget:self selector:@selector(onDisplayLink)];
-  [self.displayLink addToRunLoop:NSRunLoop.currentRunLoop forMode:NSRunLoopCommonModes];
-  __block CADisplayLink* currentDisplayLink = self.displayLink;
+  // Invalidate old vsync client if old animation is not completed.
+  [self invalidateKeyboardAnimationVsyncClient];
+  [self setupKeyboardAnimationVsyncClient];
+  __block VSyncClient* currentVsyncClient = _keyboardAnimationVsyncClient;
 
   [UIView animateWithDuration:duration
       animations:^{
         // Set end value.
         [self keyboardAnimationView].frame = CGRectMake(0, self.targetViewInsetBottom, 0, 0);
       }
       completion:^(BOOL finished) {
-        if (self.displayLink == currentDisplayLink) {
-          // Indicates the displaylink captured by this block is the original one,which also
-          // indicates the animation has not been interrupted from its beginning. Moreover,indicates
-          // the animation is over and there is no more animation about to exectute.
-          [self invalidateDisplayLink];
+        if (_keyboardAnimationVsyncClient == currentVsyncClient) {
+          // Indicates the vsync client captured by this block is the original one, which also
+          // indicates the animation has not been interrupted from its beginning. Moreover,
+          // indicates the animation is over and there is no more to execute.
+          [self invalidateKeyboardAnimationVsyncClient];
           [self removeKeyboardAnimationView];
           [self ensureViewportMetricsIsCorrect];
         }
       }];
 }
 
-- (void)invalidateDisplayLink {
-  [self.displayLink invalidate];
+- (void)setupKeyboardAnimationVsyncClient {
+  auto callback = [self](std::unique_ptr<flutter::FrameTimingsRecorder> recorder) {
+    if ([self keyboardAnimationView].superview == nil) {
```

Comment:
`self` could be deleted by the time this is executed.  There is no guarantee that it is still in memory.

Reply:
Done.

Reply:
No, I meant between `setupKeyboardAnimationVsyncClient` and `callback` executing `self` could be deleted, which would cause this code to crash.

Reply:
Done.

Reply:
Here is the scenario I'm talking about.  The way the code is setup, the `_keyboardAnimationVsyncClient` has a weak reference to the FlutterViewController.  In `-[FlutterViewController dealloc]` there is nothing that clears out that reference so the `_keyboardAnimationVsyncClient` could end up talking to the dead reference.  You may be able to fix the problem by invalidating the vsyncclient in the dealloc.  You'd have to look to make sure that would work.

Reply:
Oh,I got it. I will make sure this.

Reply:
Just like before, there is a invalidate method in `viewDidDisappear` to invalidate the vsync client. Emm I think it should be vaild.....
```objective-c

- (void)viewDidDisappear:(BOOL)animated {
  TRACE_EVENT0("flutter", "viewDidDisappear");
  if ([_engine.get() viewController] == self) {
    [self invalidateKeyboardAnimationVsyncClient];
    [self ensureViewportMetricsIsCorrect];
    [self surfaceUpdated:NO];

```

Reply:
I have added a `invalidateKeyboardAnimationVsyncClient ` method in `FlutterViewController`s `dealloc` method.

And I test this `invalidateKeyboardAnimationVsyncClient` in dealloc is valid and will be invoke correctly by deleting the `invalidateKeyboardAnimationVsyncClient` in UIView.animate(completion) in `startKeyboardAnimation` and deleting the method in `viewDidDisappear` to make sure the `invalidateKeyboardAnimationVsyncClient` in dealloc works fine and `dealloc` method of `FlutterViewController` will be invoke correctly.
![20220805115034](https://user-images.githubusercontent.com/49340347/182997884-c0bd44d5-df23-4eb4-8ddc-169bcd2e421b.jpg)


So now it can ensure the callback of `keyboardVsyncClient` will not get a dead reference.
I think just add the `invalidateKeyboardAnimationVsyncClient` in `viewDidDisappear` is enough before. But maybe there is  a  kind of scenario that `UIViewController` will be dealloc by not calling `viewDidDisappear`...
Now should be ok...😄